pilot/pkg/status/resourcelock.go

```diff
@@ -14,47 +14,189 @@
+		GroupVersionResource: i.GroupVersionResource,
+		Namespace:            i.Namespace,
+		Name:                 i.Name,
 	}
-	return false
 }
 
-func (r *ResourceLock) retrieveMutex(i interface{}) *sync.Mutex {
-	if !r.init() {
-		r.masterLock.RLock()
-		if result, ok := r.listing[i]; ok {
-			r.masterLock.RUnlock()
-			return result
+type WorkQueue struct {
+	// tasks which are not currently executing but need to run
+	tasks []lockResource
+	// a lock to govern access to data in the cache
+	lock sync.Mutex
+	// for each task, a cacheEntry which can be updated before the task is run so that execution will have latest values
+	cache map[lockResource]cacheEntry
+
+	OnPush func()
+}
+
+func (wq *WorkQueue) Push(target Resource, progress Progress) {
+	wq.lock.Lock()
+	key := convert(target)
+	_, inqueue := wq.cache[key]
+	wq.cache[key] = cacheEntry{
+		cacheVal:      &target,
+		cacheProgress: &progress,
+	}
+	if !inqueue {
+		wq.tasks = append(wq.tasks, key)
+	}
+	wq.lock.Unlock()
+	if wq.OnPush != nil {
+		wq.OnPush()
+	}
+}
+
+// Pop returns the first item in the queue not in exclusion, along with it's latest progress
+func (wq *WorkQueue) Pop(exclusion map[lockResource]struct{}) (target *Resource, progress *Progress) {
+	wq.lock.Lock()
+	defer wq.lock.Unlock()
+	for i := len(wq.tasks) - 1; i >= 0; i-- {
+		if _, ok := exclusion[wq.tasks[i]]; !ok {
+			// remove from tasks
+			t, ok := wq.cache[wq.tasks[i]]
+			wq.tasks = append(wq.tasks[:i], wq.tasks[i+1:]...)
+			if !ok {
+				return nil, nil
+			}
+			return t.cacheVal, t.cacheProgress
 		}
-		// transition to write lock
-		r.masterLock.RUnlock()
 	}
-	r.masterLock.Lock()
-	defer r.masterLock.Unlock()
-	r.listing[i] = &sync.Mutex{}
-	return r.listing[i]
+	return nil, nil
+}
+
+func (wq *WorkQueue) Length() int {
+	wq.lock.Lock()
+	defer wq.lock.Unlock()
+	return len(wq.tasks)
+}
+
+func (wq *WorkQueue) Delete(target *Resource) {
+	wq.lock.Lock()
+	defer wq.lock.Unlock()
+	delete(wq.cache, convert(*target))
+}
+
+type WorkerPool struct {
+	q WorkQueue
+	// indicates the queue is closing
+	closing bool
+	// the function which will be run for each task in queue
+	work func(*Resource, *Progress)
+	// current worker routine count
+	workerCount uint
+	// maximum worker routine count
+	maxWorkers       uint
+	currentlyWorking map[lockResource]struct{}
+	lock             sync.Mutex
+}
+
+func NewWorkerPool(work func(*Resource, *Progress), maxWorkers uint) WorkerQueue {
+	return &WorkerPool{
+		work:             work,
+		maxWorkers:       maxWorkers,
+		currentlyWorking: make(map[lockResource]struct{}),
+		q: WorkQueue{
+			tasks:  make([]lockResource, 0),
+			cache:  make(map[lockResource]cacheEntry),
+			OnPush: nil,
+		},
+	}
+}
+
+func (wp *WorkerPool) Delete(target Resource) {
+	wp.q.Delete(&target)
+}
+
+func (wp *WorkerPool) Push(target Resource, progress Progress) {
+	wp.q.Push(target, progress)
+	wp.maybeAddWorker()
+}
+
+func (wp *WorkerPool) Run(ctx context.Context) {
+	go func() {
```

Comment:
nit: it would be more efficient to have the thing checking `closing` to just check if the context is done. Then we save 1 goroutine that is just waiting here forever